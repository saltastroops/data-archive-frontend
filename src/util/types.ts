/**
 * Auth provider types.
 */
export type AuthProvider = "SSDA" | "SDB";

export interface IUser {
  authProvider: AuthProvider;
  familyName: string;
  givenName: string;
  isAdmin: boolean;
}
