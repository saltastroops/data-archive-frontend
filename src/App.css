.dimScreen {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  height: 100%;
  width: 100%;
  margin: 0;
  padding: 0;
  background: #000000;
  opacity: 0.5;
  filter: alpha(opacity=30);
  -moz-opacity: 0.3;
  z-index: 101;
}
.signal {
  border: 15px solid #fff;
  border-radius: 30px;
  height: 130px;
  left: 50%;
  margin: -15px 0 0 -15px;
  opacity: 0;
  position: absolute;
  top: 50%;
  width: 130px;

  animation: pulsate 1s ease-out;
  animation-iteration-count: infinite;
}

@keyframes pulsate {
  0% {
    transform: scale(0.1);
    opacity: 0;
  }
  50% {
    opacity: 2;
  }
  100% {
    transform: scale(3);
    opacity: 0;
  }
}

.search-results.table {
  margin-bottom: 20px;
}

.search-results.cell {
  display: flex;
  align-items: center;
  padding-left: 5px;
  white-space: nowrap;
}

.search-results.cell.header {
  background-color: white;
  border-bottom: 1px solid lightgray;
  cursor: pointer;
}

.search-results.cell.header.cart {
  cursor: text;
}

.search-results.cell span {
  overflow: hidden;
  text-overflow: ellipsis;
}

.search-results.observation-header {
  background-color: #f2fffb;
  border-bottom: 1px solid lightgray;
  border-top: 2px solid lightgray;
}

.search-results.file-row-even {
  background-color: rgb(245, 245, 245);
}

.search-results.file-row-odd {
  background-color: white;
}

.ReactVirtualized__Grid {
  outline: none;
}
