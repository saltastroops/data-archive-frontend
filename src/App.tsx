import * as React from "react";
import { Query } from "react-apollo";
import { Redirect, Route, Switch } from "react-router-dom";
import DataRequestsForm from "./components/dataRequest/DataRequestsForm";
import LoginForm, { ILoginFormCache } from "./components/LoginForm";
import NavigationBar from "./components/NavigationBar";
import RegistrationForm, {
  IRegistrationFormCache
} from "./components/RegistrationForm";
import ISearchFormCache from "./components/searchFormComponents/ISearchFormCache";
import SearchPage from "./components/searchFormComponents/SearchPage";
import UserUpdateForm, {
  IUserUpdateFormCache
} from "./components/UserUpdateForm";
import { USER_QUERY } from "./graphql/Query";

interface IUser {
  familyName: string;
  givenName: string;
  isAdmin: boolean;
}

interface IProtectedRouteProps {
  component: any;
  user: IUser | null | undefined;
  [propName: string]: any;
}

interface ICache {
  loginForm: ILoginFormCache;
  registrationForm: IRegistrationFormCache;
  searchForm: ISearchFormCache;
  userUpdateForm: IUserUpdateFormCache;
}

/**
 * A route which redirects to the login page if the user is not logged in.
 */
function ProtectedRoute({
  component: Component,
  user,
  ...rest
}: IProtectedRouteProps) {
  return (
    <Route
      {...rest}
      render={props =>
        user ? (
          <Component {...props} />
        ) : (
          <Redirect
            to={{
              pathname: "/login",
              state: { from: props.location }
            }}
          />
        )
      }
    />
  );
}

interface IAppState {
  user?: IUser;
  screenDimensions: { innerHeight: number; innerWidth: number };
}

/**
 * The data archive.
 */
class App extends React.Component<{}, IAppState> {
  state = {
    screenDimensions: {
      innerHeight: window.innerHeight,
      innerWidth: window.innerWidth
    },
    user: undefined
  };

  private cache: ICache = {
    loginForm: {},
    registrationForm: {},
    searchForm: {},
    userUpdateForm: {}
  };

  public render() {
    return (
      <Query query={USER_QUERY}>
        {({ data, loading }: any) => {
          if (loading) {
            return <p>Loading...</p>;
          }

          const currentUser =
            data && data.user
              ? {
                  familyName: data.user.familyName,
                  givenName: data.user.givenName,
                  isAdmin: data.user.roles.some(
                    (role: string) => role === "ADMIN"
                  )
                }
              : null;

          return (
            <>
              <NavigationBar user={currentUser} />

              <Switch>
                {/* search page */}
                <Route
                  exact={true}
                  path="/"
                  render={() => (
                    <SearchPage
                      cache={this.cache.searchForm}
                      screenDimensions={this.state.screenDimensions}
                    />
                  )}
                />

                {/* registration page */}
                <Route
                  exact={true}
                  path="/register"
                  render={() => (
                    <RegistrationForm cache={this.cache.registrationForm} />
                  )}
                />

                {/* login page */}
                <Route
                  exact={true}
                  path="/login"
                  render={() => <LoginForm cache={this.cache.loginForm} />}
                />

                {/* account details page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/account"
                  component={() => (
                    <h1 className="title">
                      {currentUser
                        ? `${currentUser.givenName} ${
                            currentUser.familyName
                          } Account`
                        : "No User"}
                    </h1>
                  )}
                />

                {/* data requests page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/data-requests"
                  component={() => <DataRequestsForm />}
                />

                {/* cart page */}
                <Route
                  exact={true}
                  path="/cart"
                  component={() => <h1 className="title">Cart page</h1>}
                />

                {/* admin page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/admin"
                  component={() => <h1 className="title">Admin page</h1>}
                />

                {/* update user page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/user-update"
                  component={() => (
                    <UserUpdateForm cache={this.cache.userUpdateForm} />
                  )}
                />

                {/* page not found */}
                <Route
                  component={() => <h1 className="title">Page not found</h1>}
                />
              </Switch>
            </>
          );
        }}
      </Query>
    );
  }
}

export default App;
