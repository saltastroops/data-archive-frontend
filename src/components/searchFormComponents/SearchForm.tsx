import * as _ from "lodash";
import * as React from "react";
import {
  CalibrationType,
  IGeneral,
  ISearchFormState,
  ITarget,
  ITelescope
} from "../../utils/ObservationQueryParameters";
import { TargetType } from "../../utils/TargetType";
import { isError } from "../../utils/validators";
import {
  ButtonGrid,
  DataGrid,
  ParentGrid,
  ProposalGrid,
  Spinner,
  TargetGrid,
  TelescopeGrid
} from "../basicComponents/Grids";
import DataForm from "./DataForm";
import ISearchFormCache from "./ISearchFormCache";
import ProposalForm, { validatedProposal } from "./ProposalForm";
import TargetForm, { validatedTarget } from "./TargetForm";
import TelescopeForm, { validatedTelescope } from "./TelescopeForm";

/**
 * Properties for the search form.
 *
 * cache:
 *     The cache for storing the form content.
 * search:
 *     The function for carrying out the search. It must expect an object with
 *     the general, target and telescope data as its only argument.
 */
interface ISearchFormProps {
  cache?: ISearchFormCache;
  search: ({
    general,
    target,
    telescope
  }: {
    general: IGeneral;
    target: ITarget;
    telescope: ITelescope | undefined;
  }) => void;
}

/**
 * A form for defining search parameters for an observation search, and for
 * initiating the search.
 */
class SearchForm extends React.Component<ISearchFormProps, ISearchFormState> {
  public state: ISearchFormState = {
    general: { calibrations: new Set<CalibrationType>(), errors: {} },
    loading: false,
    target: {
      errors: {},
      resolver: "Simbad",
      searchConeRadius: "",
      searchConeRadiusUnits: "arcseconds",
      targetTypes: new Set<TargetType>()
    }
  };

  /**
   * Populate the state from cached values.
   */
  componentDidMount() {
    this.setState(() => (this.props.cache as any) || {});
  }

  /**
   * Handle changes of telescope-related parameters.
   */
  public telescopeChange = (value: ITelescope) => {
    const newState = {
      ...this.state,
      telescope: {
        ...value
      }
    };
    this.updateState(newState);
  };

  /**
   * Handle changes of target-related parameters.
   */
  public targetChange = (value: ITarget) => {
    const newState = {
      ...this.state,
      target: {
        ...value
      }
    };
    this.updateState(newState);
  };

  /**
   * Handle changes of general parameters.
   */
  public generalChange = (value: IGeneral) => {
    const newState = {
      ...this.state,
      general: {
        ...value
      }
    };
    this.updateState(newState);
  };

  public render() {
    const { target, general, telescope, loading } = this.state;

    return (
      <>
        {loading && (
          <div className="dimScreen">
            <Spinner />
          </div>
        )}
        <ParentGrid>
          <TargetGrid>
            <TargetForm target={target} onChange={this.targetChange} />
          </TargetGrid>
          <ProposalGrid>
            <ProposalForm proposal={general} onChange={this.generalChange} />
          </ProposalGrid>
          <TelescopeGrid>
            <TelescopeForm
              telescope={telescope}
              onChange={this.telescopeChange}
            />
          </TelescopeGrid>
          <DataGrid>
            <DataForm general={general} onChange={this.generalChange} />
          </DataGrid>
          <ButtonGrid>
            <button
              disabled={loading}
              className="button is-primary"
              data-test="search-button"
              type="button"
              value="Search"
              onClick={this.onSubmit}
            >
              search
            </button>
          </ButtonGrid>
        </ParentGrid>
      </>
    );
  }

  /**
   * Initiate the search.
   */
  private onSubmit = async () => {
    // Add errors to the search parameter details
    const target = await validatedTarget(this.state.target);
    const general = await validatedProposal(this.state.general);
    const telescope = await validatedTelescope(this.state.telescope);

    this.updateState({
      ...this.state,
      general,
      loading: false,
      target,
      telescope
    });
    if (
      !isError(
        general.errors,
        target.errors,
        (telescope && telescope.errors) || {}
      )
    ) {
      this.props.search({ general, target, telescope });
    }
  };

  /**
   * Update the form state and the cache.
   */
  private updateState = (update: object) => {
    this.setState(
      () => update,
      () => {
        if (this.props.cache) {
          this.props.cache.general = _.cloneDeep(this.state.general);
          this.props.cache.target = _.cloneDeep(this.state.target);
          this.props.cache.telescope = _.cloneDeep(this.state.telescope);
        }
      }
    );
  };
}

export default SearchForm;
