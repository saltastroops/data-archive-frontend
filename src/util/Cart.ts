import gql from "graphql-tag";

/**
 * The cart of requested files.
 */
export class Cart {
  private cartFiles: ICartFile[];

  constructor(files: ICartFile[]) {
    this.cartFiles = files;
  }

  /**
   * The files in the cart.
   *
   * While this method returns the original array of files rather than a copy,
   * you are strongly discouraged from modifying it. Use the add abd remove
   * methods instead.
   */
  public get files() {
    return this.cartFiles;
  }

  /**
   * The number of files in the cart.
   */
  public get size() {
    return this.cartFiles.length;
  }

  /**
   * Check whether the cart contains the given file.
   *
   * Files are compared by their id. So if any cart file has the same id as the
   * given file, the function returns true, irrespective of any other key-value
   * pairs are equal.
   *
   * Parameters
   * ----------
   * file:
   *     The file whose existence in the cart is checked.
   *
   * Returns
   * -------
   * contains:
   *     Whether the cart contains the given file.
   */
  public contains(file: ICartFile) {
    return this.cartFiles.some(f => f.id === file.id);
  }

  /**
   * Adds files to the cart.
   *
   * A file in the given array is added only if it is not in the cart already.
   * The same logic as for the contains method is used to decide whether a file
   * is in the cart.
   *
   * @param files
   */
  public add(files: ICartFile[]) {
    files.forEach(file => {
      if (!this.contains(file)) {
        this.cartFiles = [...this.cartFiles, file];
      }
    });
  }

  /**
   * Removes files from the cart.
   *
   * If a file in the given array is not in the cart already, it is just
   * ignored. The same logic as for the contains method is used to decide
   * whether a file is in the cart.
   *
   * @param files
   */
  public remove(files: ICartFile[]) {
    this.cartFiles = this.cartFiles.filter(
      file => !files.some(f => file.id === f.id)
    );
  }

  /**
   * Returns the cart files grouped by observation.
   *
   * The files are returned as a map whose keys are the observation ids. Each
   * value is the set of files for the observation which have the corresponding
   * key as their id.
   *
   * Files which are not linked to an observation are collected in a set with
   * the empty string as its key.
   *
   * For example, consider a cart containing the files
   *
   * A_Obs1, B, C_Obs1, D_Obs2, E
   *
   * where A_Obs1 and C_Obs1 are linked to observation with id Obs1, D_Obs2 is
   * linked to an observation with id Obs2 and B and E are not linked to any
   * observation. Then this method returns a map with the following content.
   *
   * {
   *   '': new Set(B, E),
   *   'Obs1': new Set(A_Obs1, C_Obs1),
   *   'Obs2': new Set(D_Obs2)
   * }
   */
  public groupByObservation() {
    const groups = new Map<string, Set<ICartFile>>();
    this.cartFiles.forEach(file => {
      const key = (file.observation && file.observation.id) || "";
      if (!groups.has(key)) {
        groups.set(key, new Set<ICartFile>());
      }
      (groups.get(key) as Set<ICartFile>).add(file);
    });

    return groups;
  }
}

/**
 * Interface for a cart file.
 *
 * Properties
 * ----------
 * id:
 *     Unique file id.
 * name:
 *     File name.
 * observation:
 *     Observation to which the file is linked.
 */
export interface ICartFile {
  id: string;
  name: string;
  observation?: IObservation | null;
}

/**
 * Interface for an observation to which a cart file is linked.
 *
 * Properties
 * ----------
 * id:
 *     Unique observation id.
 * name:
 *     Observation name.
 */

interface IObservation {
  id: string;
  name: string;
}

export const CART_QUERY = gql`
  query CART_QUERY {
    cart @client {
      id
      name
      observation
    }
  }
`;

export const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($files: [CartFileInput!]!) {
    addToCart(files: $files) @client
  }
`;

export const REMOVE_FROM_CART_MUTATION = gql`
  mutation REMOVE_FROM_CART_MUTATION($files: [CartFileInput!]!) {
    removeFromCart(files: $files) @client
  }
`;
