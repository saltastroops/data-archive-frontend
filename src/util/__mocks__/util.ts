import { IUser } from "../types";

export const mockUser = ({
  authProvider,
  familyName,
  givenName,
  isAdmin
}: IUser) => ({
  authProvider,
  familyName,
  givenName,
  isAdmin
});
