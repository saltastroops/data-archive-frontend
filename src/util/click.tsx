import { ReactWrapper } from "enzyme";
import * as React from "react";

/**
 * Call the onClick method of the element wrapped in a wrapper with a mouse
 * click event.
 *
 * This function is intended for use in unit tests only. It is required because
 * the standard  Enzyme simulate does not seem to work for routing (see
 * https://github.com/ReactTraining/react-router/issues/4337).
 *
 * Properties
 * ----------
 * wrapper:
 *     Enzyme wrapper, which should wrap exactly one element.
 */
export default function click(
  wrapper: ReactWrapper<any, React.Component, any>
) {
  wrapper.props().onClick(new MouseEvent("click"));
}
