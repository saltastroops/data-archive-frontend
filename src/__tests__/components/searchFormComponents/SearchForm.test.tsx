import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import { MockedProvider } from "react-apollo/test-utils";
import { MemoryRouter } from "react-router";

import wait from "waait";
import App from "../../../App";
import SearchForm from "../../../components/searchFormComponents/SearchForm";
import { CART_QUERY } from "../../../util/Cart";
import click from "../../../util/click";

window.matchMedia = jest.fn().mockImplementation(query => {
  return {
    addListener: jest.fn(),
    matches: false,
    media: query,
    onchange: null,
    removeListener: jest.fn()
  };
});

const mocks = [
  {
    request: {
      query: CART_QUERY,
      variables: {}
    },
    result: {
      data: {
        cart: []
      }
    }
  }
];

const screenDimensions = { innerHeight: 1000, innerWidth: 1400 };

describe("Search Form", () => {
  it("should render", () => {
    const wrapper = mount(<SearchForm search={jest.fn()} />);
    expect(wrapper).toBeDefined();
  });

  it("should always render a div of class name grid-container", () => {
    const wrapper = mount(<SearchForm search={jest.fn()} />);
    const mainDiv = wrapper.find("div.grid-container");
    expect(mainDiv.length).toBeGreaterThan(0);
  });

  it("should contain five grid-item and one button", () => {
    const wrapper = mount(<SearchForm search={jest.fn()} />);
    const items = wrapper.find("div.grid-item");
    expect(items.length).toEqual(5);
    const button = wrapper.find("button.is-primary");
    expect(button.length).toEqual(1);
  });

  it("should contain target, proposal, telescope and data form", () => {
    const wrapper = mount(<SearchForm search={jest.fn()} />);
    const target = wrapper.find("div.target-form");
    expect(target.length).toEqual(1);
    const proposal = wrapper.find("div.proposal-form");
    expect(proposal.length).toEqual(1);
    const telescope = wrapper.find("div.telescope-form");
    expect(telescope.length).toEqual(1);
    const data = wrapper.find("div.data-form");
    expect(data.length).toEqual(1);
  });

  it("should update state when typing", () => {
    let value: any;

    const wrapper = mount(<SearchForm search={jest.fn()} />);
    const targetInput = wrapper.find('input[data-test="target-name-input"]');
    const targetName = targetInput.find("input");

    value = "apple";
    targetName.simulate("change", { target: { value, name: "name" } });
    expect((wrapper.find("SearchForm").state() as any).target.name).toEqual(
      "apple"
    );

    value = "";
    targetName.simulate("change", { target: { value, name: "name" } });
    expect((wrapper.find("SearchForm").state() as any).target.name).toEqual("");

    value = undefined;
    targetName.simulate("change", { target: { value, name: "name" } });
    expect((wrapper.find("SearchForm").state() as any).target.name).toEqual("");

    value = null;
    targetName.simulate("change", { target: { value, name: "name" } });
    expect((wrapper.find("SearchForm").state() as any).target.name).toEqual("");

    value = "\\ \t";
    targetName.simulate("change", { target: { value, name: "name" } });
    expect((wrapper.find("SearchForm").state() as any).target.name).toEqual(
      "\\ \t"
    );
  });

  it("should update state when a selection is changed", () => {
    let value: any;
    const wrapper = mount(<SearchForm search={jest.fn()} />);
    const resolverSelect = wrapper.find('select[data-test="resolver-select"]');
    const resolver = resolverSelect.find("select");
    value = "NED";
    const event = { target: { value, name: "resolver" } };
    resolver.simulate("change", event);
    expect((wrapper.state() as any).target.resolver).toEqual("NED");
  });

  it("should cache values and errors", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <MemoryRouter>
          <App />
        </MemoryRouter>
      </MockedProvider>
    );

    // Wait for the user query to finish
    await wait(0);
    wrapper.update();

    // Navigate to the search form
    const searchFormLink = wrapper.find('a[href="/"]').first();
    click(searchFormLink);

    await wait(0);
    wrapper.update();

    // Fill in invalid target and general information
    const searchConeRadiusInput = wrapper.find(
      'input[data-test="search-cone-radius-input"]'
    );
    searchConeRadiusInput.simulate("change", {
      target: { value: "invalid radius", name: "searchConeRadius" }
    });
    const observationNightInput = wrapper.find(
      'input[data-test="observation-night-input"]'
    );
    observationNightInput.simulate("change", {
      target: { value: "invalid date", name: "observationNight" }
    });

    // Submit the form
    const submitButton = wrapper.find('button[data-test="search-button"]');
    submitButton.simulate("click");

    await wait(0);
    wrapper.update();

    // The values have been stored in the state, and there are errors
    const searchFormState: any = wrapper.find("SearchForm").state();
    const radiusValue = searchFormState.target.searchConeRadius;
    expect(radiusValue).toEqual("invalid radius");
    const radiusErrorMessage = searchFormState.target.errors.searchConeRadius;
    expect(radiusErrorMessage.length).toBeGreaterThan(0);
    const observationNightValue = searchFormState.general.observationNight;
    expect(observationNightValue).toEqual("invalid date");
    const observationNightErrorMessage =
      searchFormState.general.errors.observationNight;
    expect(observationNightErrorMessage.length).toBeGreaterThan(0);

    // Navigate away from the search form
    const cartLink = wrapper.find('a[href="/cart"]').first();
    click(cartLink);

    await wait(0);
    wrapper.update();

    // No search form any longer
    expect(wrapper.find("SearchForm").length).toBe(0);

    // Navigate back to the search form
    click(searchFormLink);

    await wait(0);
    wrapper.update();

    // The values and errors have been re-inserted.
    const newSearchFormState: any = wrapper.find("SearchForm").state();
    expect(newSearchFormState.target.searchConeRadius).toEqual(radiusValue);
    expect(newSearchFormState.target.errors.searchConeRadius).toEqual(
      radiusErrorMessage
    );
    expect(newSearchFormState.general.observationNight).toEqual(
      observationNightValue
    );
    expect(newSearchFormState.general.errors.observationNight).toEqual(
      observationNightErrorMessage
    );
  });

  it("render correctly", () => {
    const wrapper = shallow(
      <MockedProvider mocks={mocks}>
        <SearchForm search={jest.fn()} />
      </MockedProvider>
    );
    expect(toJson(wrapper.find("SearchForm"))).toMatchSnapshot();
  });
});
