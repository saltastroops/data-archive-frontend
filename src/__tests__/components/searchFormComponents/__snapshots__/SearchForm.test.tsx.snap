// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Form render correctly 1`] = `
<SearchForm
  search={[MockFunction]}
/>
`;
